lib/membrane_http_adaptive_stream/manifest/track.ex

```diff
@@ -221,178 +437,199 @@ defmodule Membrane.HTTPAdaptiveStream.Manifest.Track do
-
-  @doc """
-  Append a partial segment to the current incomplete segment.
-
-  The current segment is supposed to be of type `:partial`, meaning that it is still in a phase
-  of gathering partial segments before being finalized into a full segment. There can only be
-  a single such segment and it must be the last one.
-  """
-  @spec add_partial_segment(t, boolean, segment_duration_t, segment_byte_size_t) ::
+  # Append a partial segment to the current incomplete segment.
+  # The current segment is supposed to be of type `:partial`, meaning that it is still in a phase
+  # of gathering partial segments before being finalized into a full segment. There can only be
+  # a single such segment and it must be the last one.
+  @spec add_partial_segment(
+          t,
+          segment_payload_t(),
+          boolean,
+          segment_duration_t,
+          segment_byte_size_t
+        ) ::
           {Changeset.t(), t()}
-  def add_partial_segment(
-        %__MODULE__{finished?: false} = track,
-        independent?,
-        duration,
-        byte_size
-      ) do
+  defp add_partial_segment(
+         %__MODULE__{finished?: false} = track,
+         payload,
+         independent?,
+         duration,
+         byte_size
+       ) do
     use Ratio
 
     partial_segment = %{
       independent?: independent?,
       duration: duration,
-      byte_size: byte_size
+      byte_size: byte_size,
+      payload: payload
     }
 
     {%Segment{type: :partial} = last_segment, segments} = Qex.pop_back!(track.segments)
 
-    metadata = %{
-      byte_offset: Enum.map(last_segment.parts, & &1.byte_size) |> Enum.sum(),
-      duration: duration,
-      independent: independent?
-    }
+    {partial_sequence_number, track} = get_partial_sequence_number(track, true)
 
     last_segment = %Segment{
       last_segment
       | parts: last_segment.parts ++ [partial_segment]
     }
 
+    new_partial_segment = %Changeset.Segment{
+      type: :partial_segment,
+      duration: duration,
+      sequence_number: partial_sequence_number,
+      byte_offset: Enum.map(last_segment.parts, & &1.byte_size) |> Enum.sum(),
+      independent?: independent?,
+      name: last_segment.name,
+      payload: payload
+    }
+
     changeset = %Changeset{
-      to_add: {:partial_segment, last_segment.name, metadata},
+      to_add: [new_partial_segment],
       to_remove: []
     }
 
     {changeset, Map.replace(track, :segments, Qex.push(segments, last_segment))}
   end
 
-  @doc """
-  Finalize current segment.
+  # Helper function for `new_segment/3`. Decides if agregated partial segments are enough to finish current segment and start new one.
+  defp should_finalize_segment?(
+         independent?,
+         duration,
+         partials,
+         partials_duration,
+         min_segment_duration
+       ) do
+    partials != [] and independent? and duration + partials_duration > min_segment_duration
+  end
```

Comment:
I'd put that to `new_segment` since it's only used there, like
```elixir
should_finalize_segment? = partials != [] and independent? and duration + partials_duration > min_segment_duration
```